pytorch_memlab/line_profiler.py

```diff
@@ -58,128 +204,143 @@ def add_function(self, func):
-                torch.cuda.empty_cache()
-            if event == 'return':
-                lineno = max(line_stat.keys()) + 1
+                self._raw_line_records.append({
+                    'code_hash': code_hash, 
+                    'line': code_info['prev_line'],
+                    'prev_record_idx': code_info['prev_record'],
+                    **torch.cuda.memory_stats()})
+                self._reset_cuda_stats()
+
+            if event == 'line':
+                code_info['prev_line'] = frame.f_lineno
+                code_info['prev_record'] = len(self._raw_line_records)-1
+            elif event == 'return':
+                code_info['prev_line'] = code_info['first_line']
+                code_info['prev_record'] = -1
             else:
-                lineno = frame.f_lineno
-            last_lineno = self.code_map[frame.f_code]['last_lineno']
-            line_stat[last_lineno].append((allocated_memory, cached_memory))
-            self.code_map[frame.f_code]['last_lineno'] = lineno
-        return
-
-    def print_stats(self, stream=None):
-        """Print the stat of each functions
+                assert False
+
+    def line_records(self, func=None, columns=DEFAULT_COLUMNS):
+        """Returns a (line, statistic)-indexed dataframe of memory stats.
+        
+        The columns are explained in the PyTorch documentation:
+        
+        https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_stats
         """
-        for code, stat in self.code_map.items():
-            show_func(
-                filename=code.co_filename,
-                trace_stat=stat,
-                stream=stream
-            )
+        if len(self._raw_line_records) == 0:
+            return pd.DataFrame(index=pd.MultiIndex.from_product([[], []]), columns=columns)
 
-    def print_func_stats(self, func, stream=None):
-        """Print the stat of a registered function"""
-        code = func.__code__
-        if code in self.code_map:
-            show_func(
-                filename=code.co_filename,
-                trace_stat=self.code_map[code],
-                stream=stream
-            )
+        line_records = _line_records(self._raw_line_records, self._code_infos)
+        return _subset_line_records(line_records, func, columns)
 
+    def display(self, func=None, columns=DEFAULT_COLUMNS):
+        """Returns an object that'll display the recorded stats in the IPython console.
 
-global_line_profiler = LineProfiler()
-global_line_profiler.enable()
+        The columns are explained in the PyTorch documentation:
+        
+        https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_stats
+        
+        To work, this needs to be the last thing returned in the IPython statement or cell.
+        """ 
+        return RecordsDisplay(self.line_records(func, columns), self._code_infos)
 
+    def print_stats(self, func=None, columns=DEFAULT_COLUMNS, stream=sys.stdout):
```

Comment:
I think we can seamlessly merge `print_stats` and `print_func_stats`.

Reply:
done